lens/lens.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package lens
+
+import (
+	"github.com/sourcenetwork/immutable/enumerable"
+
+	"github.com/sourcenetwork/defradb/client"
+)
+
+type schemaVersionID = string
+
+// LensDoc represents a document that will be sent to/from a Lens.
+type LensDoc = map[string]any
+
+type lensInput struct {
+	SchemaVersionID schemaVersionID
+	Doc             LensDoc
+}
+
+// Lens migrate items fed in to the target schema version.
+//
+// Source documents may be of various schema versions, and may need to be migrated across multiple
+// versions.  As the input versions are unknown until enumerated, the migration pipeline is constructed
+// lazily, as new source schema versions are discovered.  If a migration does not exist for a schema
+// version, the document will be passed on to the next stage.
+type Lens interface {
+	enumerable.Enumerable[LensDoc]
+
+	// Put feeds the given document into the Lens, so that its transformed output may be yielded.
+	Put(schemaVersionID schemaVersionID, value LensDoc) error
+}
+
+type lens struct {
+	lensRegistry client.LensRegistry
+
+	lensPipesBySchemaVersionIDs      map[schemaVersionID]enumerable.Concatenation[LensDoc]
+	lensInputPipesBySchemaVersionIDs map[schemaVersionID]enumerable.Queue[LensDoc]
```

Comment:
Ahhh, I mean their usage is quite heavily documented a few lines down.  And why an `input` pipe would need to be a writable queue instead of a read-only enumerable shouldn't be too confusing. Will see if I can add some comments that add value to this though.

- [ ] Add comment to pipe props